File: recursion/merge_sort.cpp
#include <iostream>
#include <vector>
using namespace std;

vector<int> merge(const vector<int> left, const vector<int> right) {
    int i = 0, j = 0, k = 0;
    int m = left.size() - 1, n = right.size() - 1;

    vector<int> arr(m + n);
    for (i = 0; i < (m + n) && j < m && k < n; i++) {
        if (left[j] <= right[k])
            arr[i] = left[i++];
        else
            arr[i] = right[k++];        
    }

    while (i < (m + n) && j < m) arr[i++] = left[j++];

    while (i < (m + n) && k < n) arr[i++] = right[k++];

    return arr;
};

vector<int> mergeSort(const int start, const int end, vector<int> v) {
    // Base condition:
    const int size = end - start;
    if (size <= 1) return v;
    if (size == 2) {
        if (v[0] > v[1]) swap(v[0], v[1]);
        return v;
    }

    // Recursion:
    const int mid = (start + end) / 2;
    vector<int> left = mergeSort(start, mid, v);
    vector<int> right = mergeSort(mid+1, end, v);

    return merge(left, right);
};

int main() {
    vector<int> v{5, 4, 3, 2, 1};
    const auto result = mergeSort(0, v.size() - 1, v);
    for (const auto n: result) cout << n << " ";
    cout << endl;
}

File: recursion/merge_sort.py
from typing import List


def merge(left: List, right: List) -> List:
    n = len(left)
    m = len(right)
    arr = [None for i in range(n+m)]
    
    i = j = k = 0
    # Merge the two sorted arrays, until one of them is fully merged
    while k < (m+n) and i < n and j < m:
        if left[i] <= right[j]:
            arr[k] = left[i]
            i += 1 
        else:
            arr[k] = right[j]
            j += 1
        k += 1
        
    # Merge the remaining elements of the left array:
    while i < n and k < (m+n):
        arr[k] = left[i]
        i += 1
        k += 1
        
    # Merge the remaining elements of the right array:
    while j < m and k < (m+n):
        arr[k] = right[j]
        j += 1
        k += 1
        
    return arr
        
    

def merge_sort(start_index: int, end_index: int, arr: List) -> List:
    # Base Condition:
    _size = end_index - start_index
    if _size <= 1:
        return arr[start_index:end_index]
    if _size == 2:
        if arr[start_index] > arr[start_index+1]:
            arr[start_index], arr[start_index+1] = arr[start_index+1], arr[start_index]
        return arr[start_index:end_index]
    
    
    # Recursive Condition:
    mid = (start_index + end_index) // 2
    left = merge_sort(start_index, mid, arr)
    right = merge_sort(mid, end_index, arr)
    
    return merge(left, right)

# Test cases:
arr = [3, 5, 1, 2, 4]
print(merge_sort(0, len(arr), arr))

arr = [3, 5, 1, 2, 4, 6]
print(merge_sort(0, len(arr), arr))

arr = [5, 1, 2]
print(merge_sort(0, len(arr), arr))

arr = [5, 1]
print(merge_sort(0, len(arr), arr))

arr = [5]
print(merge_sort(0, len(arr), arr))

str_arr = ['c', 'a', 'b']
print(merge_sort(0, len(str_arr), str_arr))

str_arr = ['c', 'a', 'b', 'd']
print(merge_sort(0, len(str_arr), str_arr))